File: src/codes/hardcoded-variability/py/rfc_flyweight.py
class NotificationTemplate:
    def __init__(self, header: str, footer: str):
        self.header = header
        self.footer = footer

    def format(self, user: str, body: str) -> str:
        return f"{self.header}\n宛先: {user}\n本文: {body}\n{self.footer}"

class Notification:
    def __init__(self, user: str, message: str, template: NotificationTemplate):
        self.user = user
        self.message = message
        self.template = template

    def send(self):
        print(self.template.format(self.user, self.message))

# 利用例
shared_template = NotificationTemplate("件名: お知らせ", "-- 通知システム")
Notification("hiroshi", "メッセージ1", shared_template).send()
Notification("satoshi", "メッセージ2", shared_template).send()

File: src/codes/hardcoded-variability/ts/rfc_flyweight.ts
// Flyweight（共通部分）
class NotificationTemplate {
  constructor(public header: string, public footer: string) {}

  format(user: string, body: string): string {
    return `${this.header}\n宛先: ${user}\n本文: ${body}\n${this.footer}`;
  }
}

// Context（可変部分を持つ）
class Notification {
  constructor(
    private user: string,
    private message: string,
    private template: NotificationTemplate
  ) {}

  send() {
    console.log(this.template.format(this.user, this.message));
  }
}

// 利用例：テンプレートは共有
const sharedTemplate = new NotificationTemplate(
  "件名: お知らせ",
  "-- 通知システム"
);

new Notification("hiroshi", "メッセージ1", sharedTemplate).send();
new Notification("satoshi", "メッセージ2", sharedTemplate).send();
